docs/hello_nextflow/03_hello_containers.md

```diff
@@ -1,82 +1,442 @@
+process cowSay {
+
+    publishDir 'containers/results', mode: 'copy'
+    container 'community.wave.seqera.io/library/pip_cowsay:131d6a1b707a8e65'
+```
+
+### 3.2. Run nextflow pipelines using containers
+
+Run the script to see the container in action.
+
+```bash
+nextflow run hello_containers.nf
+```
+
+!!! NOTE
+
+    The `nextflow.config` in our current working directory contains `docker.enabled = true`, which tells Nextflow to use Docker to run processes.
+
+Without that configuration we would have to specify the `-with-docker` flag when running the script.
+
+### 3.3. Check the results
+
+You should see a new directory called `containers/results` that contains the output of the `cowsay` process.
+
+```console title="containers/results/cowsay-output-Bonjour.txt"
+  _______
+| Bonjour |
+  =======
+       \
+        \
+          ^__^
+          (oo)\_______
+          (__)\       )\/\
+              ||----w |
+              ||     ||
+```
+
+### 3.4. Explore how Nextflow launched the containerized task
+
+Let's take a look at the task directory for one of the cowsay tasks to see how Nextflow works with containers under the hood.
+
+Check your the output form your `nextflow run` command to find the task ID for the `cowsay` process.
+Then checkout the task directory for that task.
+
+```bash
+tree -a work/8c/738ac55b80e7b6170aa84a68412454
+work/8c/738ac55b80e7b6170aa84a68412454
+├── .command.begin
+├── .command.err
+├── .command.log
+├── .command.out
+├── .command.run
+├── .command.sh
+├── .exitcode
+├── cowsay-output-Bonjour.txt
+└── output-Bonjour.txt -> /workspace/gitpod/nf-training/hello-nextflow/work/0e/e96c123cb7ae9ff7b7bed1c5444009/output-Bonjour.txt
+
+1 directory, 9 files
+```
+
+Open the `.command.run` file which holds all the busywork that Nextflow does under the hood.
+
+```bash
+code work/8c/738ac55b80e7b6170aa84a68412454/.command.run
+```
+
+Search for `nxf_launch` and you should see something like this:
+
+```bash
+nxf_launch() {
+    docker run -i --cpu-shares 1024 -e "NXF_TASK_WORKDIR" -v /workspace/gitpod/nf-training/hello-nextflow/work:/workspace/gitpod/nf-training/hello-nextflow/work -w "$NXF_TASK_WORKDIR" --name $NXF_BOXID community.wave.seqera.io/library/pip_cowsay:131d6a1b707a8e65 /bin/bash -ue /workspace/gitpod/nf-training/hello-nextflow/work/8c/738ac55b80e7b6170aa84a68412454/.command.sh
+}
+```
+
+As you can see, Nextflow is using the `docker run` command to launch the task.
+It also mounts the task's working directory into the container, sets the working directory inside the container to the task's working directory, and runs our templated bash script in the `.command.sh` file.
+All the hard work we learned about in the previous sections is done for us by Nextflow!
+
+### Takeaway
+
+You know how to use containers in Nextflow to run processes.
+
+### What's next?
+
+An optional exercise to fetch quotes on computer/biology pioneers using the `quote` container and output them using the `cowsay` container.
```

Comment:
I get to this bit and I'm like ahhhh that's why he wanted me to get the quote container. Reinforces that that bit belongs hereabouts, imo.

Reply:
Agreed. Moved.